### ISSUE TITLE
[Discussion] Project Status

### ISSUE THREAD

**Issue by KevDoy** [FreeRCT/FreeRCT#149]
What's the realistic status on FreeRCT? 

I know activity was rather slow back in 2016 when I contributed some pixel art, but I am curious if we'll see the project pick up again in the future or if it's considered dormant (or worse).

Cheers,

Kev

**Comment by Alberth289346** [comment]
Who knows.

After 5 years I concluded I would have to drag the thing forward for another 3-5 years, which was just too much to do by just me.
Clearly something went wrong, if only I knew what.

I do still ponder about alternative paths every now and then, but I have currently no space in my life to actually try it.

**Comment by Noordfrees** [comment]
Hi

A while back I stumbled over FreeRCT and liked the idea a lot, being a fan of the original RCT1. I have experience coding C++ and if someone would point me to something suited for a newcomer, I could try and contribute code to the project :) 

Best regards 
Benedikt

**Comment by LordAro** [comment]
Unfortunately, the project is essentially dead. Everyone involved either got tired of it or ran out of time. You're welcome to work on what's there though, I don't particularly have any issues assigning ownership to others if that's desirable.

The real answer here is that you should go look at https://github.com/OpenRCT2/OpenRCT2. It was first announced about a year after we got going with FreeRCT and it's been extremely successful - it's quite likely hoovered up any chance of other contributors coming back to this.

**Comment by Noordfrees** [comment]
Thanks for the reply. I don't mind contributing to a project with few active developers – maybe I can help revive it :) I think I'll start by investigating the assert fail I just encountered in master, and then work on adding some features.

What I don't like about OpenRCT2 is that they provide only the engine and require the original game files. A truly free game should be modder-friendly and never depend on proprietary assets. Even if OpenRCT2 is bigger and more successful, a fully free and independent new-implementation like FreeRCT seems way more appealing to me.

**Comment by LordAro** [comment]
I imagine they'll get there eventually. OpenTTD (which has been made in much the same way) was dependent on the original TTD files for a quite some time as well. (Well, 6 years, when it's now nearly 17 years old, but still)

**Comment by Newbytee** [comment]
> What I don't like about OpenRCT2 is that they provide only the engine and require the original game files. A truly free game should be modder-friendly and never depend on proprietary assets

There's this: https://github.com/OpenRCT2/OpenGraphics Not sure how complete it is at this point.

**Comment by Alberth289346** [comment]
Some thoughts what could be done differently.

- Painting pixels on an SDL surface by the CPU is a dead path nowadays, stuff needs to move to the GPU.
- Have regular playable releases.
- Don't have pure CPP code. It's difficult and has few developers. Some mix of CPP and eg Python may have a better future due to more Python programmers being around.

**Comment by Noordfrees** [comment]
> * Painting pixels on an SDL surface by the CPU is a dead path nowadays, stuff needs to move to the GPU.

I don't know if you have anything specific in mind, but IMHO OpenGL would be a good choice. It's pretty standard nowadays and allows efficient and high-quality rendering.

> * Have regular playable releases.

:+1:

> * Don't have pure CPP code. It's difficult and has few developers. Some mix of CPP and eg Python may have a better future due to more Python programmers being around.

Personally I don't like Python much, though the argument is a good one. But if I may make a suggestion: As a long-term aim the definitions of rides and other assets could be moved to Lua scripts, which offers users the way to easily – and without knowing much programming stuff at all yet – modify the game. This also attracts new developers who start by working with Lua scripting and later eventually learn to code in C++ as well (speaking from personal FOSS experience here).


---

By the way, hoping this is not the wrong place to ask, are there some documents or links a new contributor should read? Like about code style, dos and donts, or other things one should know?

**Comment by Alberth289346** [comment]
> I don't know if you have anything specific in mind, but IMHO OpenGL would be a good choice.

Being a Linux user, OpenGL would be my first choice too, not sure how cross-platform that is.

> As a long-term aim the definitions of rides and other assets could be moved to Lua scripts

We may have different views on what should be done in CPP and what should be done in the scripting language.

I dabble in CorsixTH as well, which is a little C++ and a big pile of Lua. In my experience there, indeed, it's simple to write, but Lua is a too small language, it doesn't scale well. Useful if you write simple short programs, but if want to write 3/4 of the program in it, it's going to be a mess with global variables by default, no standard class definition (everyone defines their own class system first), starting at 1 while cpp starts at 0, value 0 not being false, and no way to distinguish between an empty variable and a non-existent variable. Also, Lua tools either don't exist or barely work, the Lua development community is very small it seems. Besides a user-manual, the only useful documentation is the Lua reference manual, which is extremely cryptic at some spots to say the least.
So this is why I stayed away from suggesting Lua here.

> By the way, hoping this is not the wrong place to ask, are there some documents or links a new contributor should read?

You mean the stuff in `/doc` ?

**Comment by Noordfrees** [comment]
> > I don't know if you have anything specific in mind, but IMHO OpenGL would be a good choice.
> 
> Being a Linux user, OpenGL would be my first choice too, not sure how cross-platform that is.

There's a bug that some graphics drivers under windows only provide OpenGL version 1.1.0. Other than that, OpenGL 2.1 has been in use under all common OSs for years in https://github.com/widelands/widelands/ with no major issues.

> > By the way, hoping this is not the wrong place to ask, are there some documents or links a new contributor should read?
> 
> You mean the stuff in `/doc` ?

:+1: thanks

**Comment by Newbytee** [comment]
There are also things like [Sokol](https://github.com/floooh/sokol) and [bgfx](https://github.com/bkaradzic/bgfx). bgfx especially can be a bit less straightforward than just writing OpenGL which isn't really ideal here, but the positive side is that you get support for a load of different graphics APIs without having to write multiple render paths.

Sokol supports GLES2/WebGL, GLES3/WebGL2, GL3.3, D3D11 and Metal and bgfx supports Direct3D 9, Direct3D 11, Direct3D 12, GNM (PS4 API), Metal, OpenGL 2.1, OpenGL 3.1+, OpenGL ES 2, OpenGL ES 3.1, Vulkan, WebGL 1.0, WebGL 2.0, WebGPU/Dawn.

**Comment by Alberth289346** [comment]
As it's all 2D, we need very little OpenGL. Mostly one or more layers that display the images at 1:1 (1 pixel image = 1 pixel screen).
Layers that may be useful

- GUI windows
- Weather (mostly rain)
- The game world, depth sorting is likely useful here.

Obviously, it can be done all on a single layer by drawing in the correct order, even without depth (as that is currently being done).

**Comment by KevDoy** [comment]
I do want to add that 1px image = 1px on screen may not be the best idea with HDPI displays. There should definitely be some scalability

**Comment by Alberth289346** [comment]
Fair enough. Scaling is the typical programmer solution. From an graphic artist point of view I can imagine that having images at different sizes is much more preferable as it gives higher quality graphics.

But yeah, something to find a path in.

**Comment by KevDoy** [comment]
If you stick to pixel art, it should scale nicely regardless 😃

**Comment by Alberth289346** [comment]
Somewhat it does, but an artist with a bigger image has more space to add details, or make some part just a little better to it's more prominent to the eye. A computer algorithm has no knowledge of what is important in an image, it treats all colours in the same way. It's easy to program (and lacking a higher resolution image likely the least bad option), but it doesn't beat a good artist.

But eh, currently we have neither :)

**Comment by Alberth289346** [comment]
Can we discuss why Python is a bad choice?
If it is about indentation, I really fail to see why not being able to make indenting mistakes is a bad choice, so please explain.

Also, what other options exist? I have run out currently.

**Comment by Noordfrees** [comment]
Perhaps before settling on a language we should discuss what exactly should be moved out from C++

If we want this for user-creatable content (scenarios, ride definitions, …) then we need a *scripting* language that can be interpreted at run-time by the engine (which remains pure C++). I suggested Lua because I have some experience with coding a C++/Lua interface (using Eris); but there are plenty of alternatives (though none I've much worked with before personally so I couldn't actually name a good choice now).
If we actually want to move parts of the engine to another language – I'm not sure this is the right approach, as then anyone who wants to do a larger change will need to know two languages rather than one.

**Comment by Alberth289346** [comment]
I expect only a handful people at best that are knowledgeable about C++ and the game code enough to make changes in C++. To bring in more developers, we need a simpler language for the bulk of the code I think.

I don't have a list, but my first rough idea is, move everything that is not low-level (ie OS facing) and not time-critical to the scripting language. Mostly window handling at OS level, and animation can't move.

For user-creatable content, all the RCD data files are for that. You can define a lot, and make your own rides, including graphics.
There is currently no scenario-like data iirc, but you could easily setup a level-file and add limitations and win/loose conditions. (Something we should have at some point anyway.)

I am not sure what you want with a scripting language in a scenario. What are useful changes you can make to eg a rollerocaster by script? A second problem is that having third-party code like that is a huge security risk. Sooner or later someone will make an awesome mod, and also install a backdoor, encrypt a few key-files, or simply delete everything at the first Friday in June.

EDIT: Biggest problem, is likely that changes here means throwing away a lot of code that already exists.

**Comment by Newbytee** [comment]
In addition to what's already been said, I think setting up a milestone for a 0.1.0 release and deciding what we want to have done for that release would be a good idea. This both would make it easier for new contributors as they know what things they can/should work on, and let us make a first release to get some publicity and possibly have it packaged.

**Comment by Newbytee** [comment]
Some ideas for what the 0.1.0 milestone might be:

 - Title screen with options to load a scenario or save, play sandbox, settings, and quit
 - One or two preloaded scenarios
 - Make FreeRCT installable and support global and user locations for things like scenarios, settings, and so on. Partly worked on in https://github.com/FreeRCT/FreeRCT/pull/123 but I doubt we'll see this guy finish it up. Still, his progress might be useful for someone interested in implementing it.
 - Fix some relevant bugs

I don't think it'd need to be more than that. We're not a lot of people so let's not overshoot, but I think these would make it feel more like game and make it more viable to package (putting it on Flathub might be a good idea, for example). It'd probably also be a good idea to have a Windows .exe in the release.

Thoughts?

**Comment by Noordfrees** [comment]
+1 for a title screen and making FRCT installable. I think scenarios might be a bit much for a first beta release, the sandbox mode would suffice for now. More urgently, imho, we could do with one or two gentle rides (that's something I'm eager to work on) and/or thrilling rides (which are basically the same thing; there's no category for those yet, is there a reason for this?).

**Comment by Alberth289346** [comment]
Oh, new rides!
I had a carousel lying around I remembered, just pushed it: https://github.com/Alberth289346/FreeRCT/tree/new-graphics
Found a ride entrance and ride exit house too, added that as well.

EDIT: The only reason why we don't have thrill rides is because nobody added it so far :)

**Comment by KevDoy** [comment]
So is anyone willing to help maintain, review code, merge prs, etc?

**Comment by Noordfrees** [comment]
I've set up a PPA with the current state-of-the-art. I'd appreciate it if someone else could test that the packages work out of the box. Packages are available for Ubuntu `Bionic Focal Groovy Hirsute Impish`. Installing is as easy as typing
```
sudo add-apt-repository ppa:nordfriese/freerct-daily
sudo apt update
sudo apt install freerct
```
Then start it with `freerct`.
Late edit: Should this fail with a segfault on startup, please try whether passing `-a nds_DE` fixes this and report :)

Tomorrow I'll make a PR with the debian rules changes that were required to get the builds working.

**Comment by Noordfrees** [comment]
Is there anything left that urgently needs fixing or implementing before we can release version 0.1? If there's no reply I'll create the release [branch](https://github.com/FreeRCT/FreeRCT/tree/release-0.1) and tag tomorrow.

**Comment by Newbytee** [comment]
It depend so on whether we want to have a Flatpak available on Flathub for the 0.1 launch. I would like to have that, but if you don't think it's important we can of course arrange that later.

**Comment by fnetX** [comment]
As announced in my latest email, I created a SocialMedia profile on Mastodon and I'll start presenting the project there. https://mastodon.technology/@FreeRCT

If anyone needs access to that account, too, I'm happy to share the credentials (or change the mail address to some @freerct.net domain so people can reset it if needed).

By the way, the website will show up as verified in the profile when this is added to the homepage: `<a rel="me" href="https://mastodon.technology/@FreeRCT">Mastodon</a>`

**Comment by Newbytee** [comment]
So what's the current plan? On the 19th we tag v0.1 and announce it as the initial release?

**Comment by Noordfrees** [comment]
Yep! Or in more detail:
- On the 19th, I create a branch named "release/0.1" and adapt its CMakeLists.txt to set the default version string to "0.1". This should probably be the only change needed but something else might turn up in the last minute.
- I tag that commit as "v0.1" and turn it into a [release](https://github.com/FreeRCT/FreeRCT/releases) on GitHub. I'll upload the release installer artifacts from the v0.1 CI build there (after unpacking the outer zip wrapper locally, for user convenience).
- The Flathub PR can be updated with the commit hash and merged. If the PR provides a standalone flatpak build (?) this should be uploaded to the GitHub release as well.
- I add links to all those files on the [homepage Download page](https://freerct.net/download#releases), and a news item on the front page.
- We announce the release to other sites – Mastodon, LibreGameWiki, and any [other relevant sites](https://github.com/FreeRCT/FreeRCT/wiki) we think of; and look into getting it packaged for more distros. [![Packaging status](https://repology.org/badge/tiny-repos/freerct.svg)](https://repology.org/project/freerct/versions)
- The master branch will need a one-liner change to set development version strings as "0.**2**~…".

**Comment by fnetX** [comment]
Is the game fully translated? It's often good to finish this before releasing. I noticed some verbose build status messages complaining about missing translations, but looking at them I can't see what's left in the langs I speak. Also, I don't fully understand the system. For example, why is the diff between en_US en en_GB so large with things mainly moved instead of changed ...

Translators are often not as tech-savy, or let's say it the other way round: If we want more contributions to this topic, we should probably look into streamlining the workflow. Maybe something to consider after the release.

**Comment by Noordfrees** [comment]
The number of missing strings per language as of 2e2b4ae4 are:
```
en_GB: source language
de_DE:   0
nds_DE:  0
sv_SE:  13
en_US: 191
nl_NL: 200
da_DK: 227
fr_FR: 234
es_ES: 238
```
We don't currently have active translators for languages other than Swedish, German, and Low German.

Our language files are rather hard to maintain because translators need to go through the translation and reference files manually to see which keys are missing. If this is in fact a standardised file format with existing tools to do this automatically I did not see a reference to it…

I do a good deal of Low German translations work for various projects on [Transifex](https://www.transifex.com/), which, although proprietary, is free to use for open-source projects, supports a large number of different file formats, and has a number of useful features like e.g. translation memory and issue management (not to mention that both-way syncs between GitHub and Transifex can be fully [automated](https://github.com/widelands/widelands/actions/workflows/i18n.yaml)). IMHO it would be worth considering to switch to a more widespread format and use Transifex for all translation work.

**Comment by fnetX** [comment]
Please avoid vendor lock in if possible. I'm also not happy about using GitHub and GH Actions already. You can also use Weblate for example.

**Comment by fnetX** [comment]
Oh, and can you tell me how you created these stats?

**Comment by Newbytee** [comment]
I think Potfiles are nice together with Poedit.

**Comment by Noordfrees** [comment]
By hand, by looking at the logs of a clean build and adding all values in a calculator 😜

I don't think we're in danger of vendor lock-in currently. Although our CI would need some rewriting if we wanted to migrate it, but Git is such a widespread VCS that we could relocate the repo anytime if we wanted to. Transifex is also just a platform and not tied to any particular format. Nothing against other providers of course – I hadn't heard of Weblate yet, will have a look at it.

I've worked a lot with gettext and its PO files. They're very widespread and there are great tools for both online and offline translation, but gettext can be a real pain to use on the coding side – at Widelands we need to ship our own copy of gettext with the source code because it's unavailable on some platforms, and its dependency on ever-changing libintl frequently mandates code updates, and it requires fragile messing with environment variables to get the desired target language. I'd prefer to use some format that we can parse without an additional 3rd-party library as we can now.

**Comment by Noordfrees** [comment]
By the way, daily builds are now deployed to the homepage automatically every night where anyone can download them without a github login:
https://freerct.net/download#daily

**Comment by Newbytee** [comment]
Regarding the Linux (Flatpak) download, are you thinking we'd host a .flatpak package that can be downloaded from freerct.net, or just link to Flathub?

**Comment by Noordfrees** [comment]
Since our release packages are quite small I wouldn't be opposed to hosting copies of all bundles on our webserver. Is there a reason against doing this with flatpak as well and/or would it be difficult to provide the standalone package? If the answer is no let's host one ourselves.

**Comment by Newbytee** [comment]
It's not impossible or necessarily difficult per se, but for FreeRCT to run you need to have the Freedesktop 21.08 Flatpak runtime installed, and you have to get that from a repository like Flathub unless we want to host it ourselves. I think it would make more sense to link Flathub from a UX perspective.

**Comment by Noordfrees** [comment]
Good point, let's link to flathub then. Could you upload a flatpakref file to the [release](https://github.com/FreeRCT/FreeRCT/releases/tag/0.1) when FreeRCT is deployed on Flathub?

**Comment by Newbytee** [comment]
Sure :+1:

**Comment by fnetX** [comment]
Seems like the website is currently down? @Noordfrees

**Comment by Noordfrees** [comment]
Restarted it, should work again. Weirdly there's nothing in the web server logs. I should definitely add some proper monitoring…

**Comment by Newbytee** [comment]
We were featured on GamingOnLinux: https://www.gamingonlinux.com/2022/03/freerct-sees-a-first-release-trying-to-capture-the-feel-of-rollercoaster-tycoon

**Comment by fnetX** [comment]
Yeah, reception on Mastodon was also quite good, I didn't expect > 60 followers in just a few days.

Probably thanks to the toot of FediFollows here: https://mastodon.online/@FediFollows/107984409532909149

**Comment by Newbytee** [comment]
> Good point, let's link to flathub then. Could you upload a flatpakref file to the [release](https://github.com/FreeRCT/FreeRCT/releases/tag/0.1) when FreeRCT is deployed on Flathub?

One thing that doesn't really make sense about this is that the Flatpakref always will point to the latest stable version, so it might be a bit confusing if someone expects that link to install the FreeRCT 0.1 Flatpak.

**Comment by Flipp3rrr** [comment]
> We don't currently have active translators for languages other than Swedish, German, and Low German.
> 
> Our language files are rather hard to maintain because translators need to go through the translation and reference files manually to see which keys are missing. If this is in fact a standardised file format with existing tools to do this automatically I did not see a reference to it…

I would be willing to help with Dutch translations, I noticed some Dutch translations tend to be unclear or only partially translated so I'll work on that too.

**Comment by Alberth289346** [comment]
Great work everyone!

**Comment by Newbytee** [comment]
I ended up uploading a .flatpak file to the release after all, so people who want it can download old versions of the Flatpak in the future if they want that for whatever reason. If I understand Flatpak's documentation correctly, it should prompt users to download the necessary runtime components from Flathub if they are missing them (I set Flathub as runtime repo when building the .flatpak bundle), so my earlier concern shouldn't be an issue.

**Comment by KevDoy** [comment]
Hey everyone,

One of the pixel artists here. I absolutely love seeing all of the progress. Incredible work, and I'm loving what I'm seeing. I'm going to take another crack at some of the icons soon as well.

Cheers!

**Comment by technicaljicama** [comment]
Hello! I wanted to thank the developers for the game having it copyright friendly etc. (very cool!).
So what i wanted to say here, is that the renderer is pretty slow on a not uptodate CPU. As you can See in my psvita port (either forum or my profile) the rendering is slow (about 4 fps at best) with overclocking to 500mhz. I would really recommend an OpenGL renderer. I might even help with it when i have More experience with OpenGL! Targeting OpenGL/es 2.0 is not a Bad choice! Since it is mostly 2D it should get the job done :). 

Thanks, PSVitaDev

**Comment by Noordfrees** [comment]
Migrating the rendering to OpenGL is definitely on my wishlist for v0.2. Unfortunately I have very little time currently, but I hope to get around to putting more effort into FreeRCT again in July and will try to tackle this then.

**Comment by Newbytee** [comment]
:+1: for going with OpenGL ES 2.0 rather than OpenGL 2.0 or such. It has way broader support on embedded platforms.

**Comment by Noordfrees** [comment]
Hi all,

GitHub is starting to roll out mandatory 2 Factor Authentication to all users over the course of this year. I am opposed to this policy and don't intend to continue using GitHub when it hits. I also believe that this requirement will deter many potential new contributors from simply reporting a bug or proposing a patch, something FreeRCT would greatly suffer from. 

Quite a number of projects are already leaving GitHub over other reasons (especially the Copilot copyright infringement debacle), and I would like to discuss moving FreeRCT to a different code hosting provider too.

There are a number of good options. I personally like Launchpad (which supports Git now and has a great bugtracker, though it doesn't have a CI of its own and its merge request UI is ugly). I also hear good things about Codeberg and GitLab.

What are your opinions on this? Should we move FreeRCT to a better hosting provider, and if yes, where to?

**Comment by Newbytee** [comment]
> Hi all,
> 
> GitHub is starting to roll out mandatory 2 Factor Authentication to all users over the course of this year. I am opposed to this policy and don't intend to continue using GitHub when it hits. I also believe that this requirement will deter many potential new contributors from simply reporting a bug or proposing a patch, something FreeRCT would greatly suffer from.
> 
> Quite a number of projects are already leaving GitHub over other reasons (especially the Copilot copyright infringement debacle), and I would like to discuss moving FreeRCT to a different code hosting provider too.
> 
> There are a number of good options. I personally like Launchpad (which supports Git now and has a great bugtracker, though it doesn't have a CI of its own and its merge request UI is ugly). I also hear good things about Codeberg and GitLab.
> 
> What are your opinions on this? Should we move FreeRCT to a better hosting provider, and if yes, where to?

I don't really understand how mandating 2FA is a bad thing, but I'm also all for moving away from GitHub, so I'll share my thoughts:

I would rather not use Launchpad. But, I'll admit that's just because I think the UI is ugly and that I don't have any experience with it. However, I can imagine I'm not the only one who has that perception of it, and I imagine it may as such result in less outside contributions. Codeberg or GitLab would be my preferred options given that I'm familiar with them and I think they are similar enough to be picked up by anyone who's use GitHub. We were considering moving postmarketOS' git repositories to Codeberg, but for us the blocker was the CI situation at the time. As I recall, the CI wasn't widely available or guaranteed to have good uptime. If this has improved, Codeberg would be my preferred option. Otherwise, GitLab.

**Comment by KevDoy** [comment]
That's disappointing. I am with @Newbytee on not understanding why 2FA is is a bad thing. I personally understand people's hangups about GitHub, however I also believe it's the best place to find and engage with collaborators with similar ideas and goals. I'll be sad if this community fragments, but also understand. Cheers.

**Comment by Alberth289346** [comment]
II would expect other hosting providers will eventually also move to 2FA, it's happening everywhere. As such trying to avoid 2FA may be a futile temporary escape.

That said, I am fine with a move, I know Gitlab, but no doubt the other options are fine too. I have no strong preferences.

**Comment by fnetX** [comment]
I am biased, as I am the executive director of Codeberg e.V., but I obviously welcome the project over at Codeberg.org, and I'm happily available to assist with the migration. We offer a hosted Woodpecker CI. I think we won't require 2FA for guests / issue reporters. A big bonus is the idea that in the future instances of Forgejo (including Codeberg) will federate with each other, so that people can report bugs and send pull requests without registering with every platform.

Still, I think that 2FA is a great idea for every developer who has elevated permissions in projects (such as updating homepage or publishing releases). Even having a TOTP software on the same device already helps against keygrabbers, for example.

**Comment by Noordfrees** [comment]
I am going to leave GitHub completely in a couple of weeks. Over at @Widelands we're now starting to use Codeberg and I like it. So, should we go ahead and migrate FreeRCT to Codeberg?


Regarding CI, as I understand it Codeberg's Woodpecker is Linux-only, is that correct @fnetX? If we don't want to lose integration for Windows and MacOS, we could also set up a simple one-way push mirror to the existing GitHub repo to retain our current GH Actions setup and daily builds.

**Comment by Newbytee** [comment]
> I am going to leave GitHub completely in a couple of weeks. Over at @widelands we're now starting to use Codeberg and I like it. So, should we go ahead and migrate FreeRCT to Codeberg?
> 
> Regarding CI, as I understand it Codeberg's Woodpecker is Linux-only, is that correct @fnetX? If we don't want to lose integration for Windows and MacOS, we could also set up a simple one-way push mirror to the existing GitHub repo to retain our current GH Actions setup and daily builds.

Moving to Codeberg is fine by me, but if Codeberg starts requiring 2FA too, will you leave Codeberg as well? I'm all for leaving GitHub, but I do think that 2FA is something that will be required by more and more services in the future and I don't see why Codeberg would be exempt from this.

**Comment by Noordfrees** [comment]
Well, I have created a migration for the repos at
https://codeberg.org/FreeRCT
https://codeberg.org/FreeRCT/FreeRCT
https://codeberg.org/FreeRCT/FreeRCT-website

---

> if Codeberg starts requiring 2FA too, will you leave Codeberg as well?

I hope they won't, but I am hopeful with regard to the quote by fnetx from above:
> A big bonus is the idea that in the future instances of Forgejo (including Codeberg) will federate with each other, so that people can report bugs and send pull requests without registering with every platform.

which seems to me to imply that even if Codeberg took a very bad turn one day one could still contribute from any other Forgejo hoster and is less dependent on one hosting provider's policy decisions.

**Comment by Newbytee** [comment]
> even if Codeberg took a very bad turn

How come you consider 2FA "a very bad turn"? 2FA helps make supply chain attacks significantly more difficult and while this isn't a big risk right now as we do not have any auto-update mechanism, the moment we do, gaining access to the controls for making new releases suddenly means you as a malicious actor can deploy code remotely on other people's computers. Granted, there are other ways that could be secured, but most — if not all — of them would likely involve something similar to 2FA. This may feel unlikely, but it has happened to much bigger names than FreeRCT who you'd hope have more resources to secure their pipelines.

If you have a reason to not want to use 2FA I'm okay with that, but I hope you understand that there are security implications of not doing so that extend beyond your own account's safety.

**Comment by Noordfrees** [comment]
2FA is in my mind a major inconvenience whose benefits I do appreciate,
but it is not the one thing with which all stands and falls here. If
GitHub had a perfect track record of being developer- and FOSS-friendly
I'd be prepared to cut them some slack and grudgingly go along with this
guideline; just as their increasingly bad policy decisions over the
years had already slowly been driving me to start considering
alternatives, and this was the (very heavy) straw to break the camel's
back. What I hate so much about the 2FA policy specifically is that it
is completely out of proportion. I have no problem using
employer-mandated 2FA at work, where there are company secrets to
protect and the time I have to spend fiddling with the authenticator
application (which quite sums up) is part of my paid work time. And if I
developed some popular library with millions of users I would accept
that this implies an increased responsibility to keep stuff secure. But
I develop a few games with a moderate user base, not high-impact
libraries, and there's no compensation for enduring the pain points of
the services used. I am not willing to tolerate too high an amount of
inconvenience for low-security-relevant projects I contribute to in my
spare time unless with a very good reason, which I don't see even
remotely warranted.

And setting up any auto-update mechanisms that don't expect independent
confirmation from the dev team sounds like asking for trouble, a repo
can *never* be kept entirely secure, it's not like stealing or
brute-forcing an admin's password is the _only_ significant attack vector…

**Comment by Newbytee** [comment]
Okay, that's fair enough. Thanks for explaining. I'm up for doing the move.

**Comment by Noordfrees** [comment]
Well, anyone who wants access to the Codeberg org[1] please tell me your
CB username so I can add you there.

I already created a new PR there to fix the WebAssembly bugs.

I have also set up a Weblate project on Codeberg Translate[2].

And I've taken the liberty of changing the website's links over to
codeberg already :)

[1] https://codeberg.org/FreeRCT
[2] https://translate.codeberg.org/projects/freerct/freerct/

**Comment by Newbytee** [comment]
I'm https://codeberg.org/Newbyte on there.

**Comment by Alberth289346** [comment]
Added a redirection to the code berg site and the website here.

**Comment by KevDoy** [comment]
Does it make sense to replace the current repo with a mirror to what is happening on codeberg? Restrict opening issues and contributions on github, but make the code available?

**Comment by Alberth289346** [comment]
I am not sure codeberg moved at all. This repo and the one at codeberg has the same number of commits in master.

**Comment by Noordfrees** [comment]
The Codeberg repo is set up with a Push mirror to GitHub to keep the
code in this repo's master branch up to date. The mirror is one-way
though. The last commit happened in November; there are a few open PRs
on Codeberg awaiting review, but the pace of development has not become
any less glacial by the move ;)

**Comment by Alberth289346** [comment]
Ok, thanks for letting me know. I wasn't aware you are pushing updates to here.
